diff --git hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/AzureNativeFileSystemStore.java hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/AzureNativeFileSystemStore.java
index 7111dcf802d..8a257163169 100644
--- hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/AzureNativeFileSystemStore.java
+++ hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azure/AzureNativeFileSystemStore.java
@@ -916,6 +916,7 @@ private void connectToAzureStorageInSecureMode(String accountName,
     rootDirectory = container.getDirectoryReference("");
 
     canCreateOrModifyContainer = true;
+    connectingUsingSAS = true;
   }
 
   /**
